api/server/router/container/container_routes.go

```diff
@@ -562,14 +563,45 @@ func (s *containerRouter) postContainersCreate(ctx context.Context, w http.Respo
 			config.Healthcheck.StartInterval = 0
 		}
 
-		for _, m := range hostConfig.Mounts {
-			if m.BindOptions != nil {
-				// Ignore ReadOnlyNonRecursive because it was added in API 1.44.
-				m.BindOptions.ReadOnlyNonRecursive = false
-				if m.BindOptions.ReadOnlyForceRecursive {
-					return errdefs.InvalidParameter(errors.New("BindOptions.ReadOnlyForceRecursive needs API v1.44 or newer"))
+		// Transform read-only binds into Mounts.
+		// This allows us to default ReadOnlyNonRecursive to true.
+		var newBinds []string
+		parser := mounts.NewParser()
```

Comment:
We already error out in case there are duplicate mountpoints: https://github.com/moby/moby/blob/9d1541526c5beab51d2344b47624a160f199bbe6/daemon/volumes.go#L179

Reply:
We should not be doing this in the router at all.
I think it would be better to add an option to toggle which behavior to use and pass that down the stack.

Reply:
Done